### PR TITLE
[nRF Connect] Port shell example

### LINKED ISSUE #2425: [nRF Connect] Port shell example
#### Problem
The shell application can be pretty useful to test various CHIP components, but is unavailable for the nRF Connect platform

 #### Proposed Solution
Port the shell example to nRF Connect SDK

### PR DESCRIPTION
**Pull request by Damian-Nordic**
#### Problem
The shell application can be pretty useful to test various CHIP components, but is unavailable for the nRF Connect platform

 #### Summary of Changes
Port the shell example to nRF Connect SDK:
1. add streamer layer for zephyr
2. add NCS project in examples/shell/nrfconnect
3. control building lib/shell via Kconfig

fixes  #2425

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

File: examples/shell/main.cpp [hunk 1]
```diff
@@ -39,10 +40,14 @@ void cmd_otcli_init();
 
 int main(void)
 {
+    // Initialize the default streamer that was linked.
+    assert(streamer_init(streamer_get()) == 0);
```
Comment (mspang) [inline examples/shell/main.cpp:44]: side effects in assert() will break in release builds

**Review by mspang — CHANGES_REQUESTED** [review]
assert() needs fixing

File: examples/shell/main.cpp (hunk 1 above)
Comment (Damian-Nordic) [reply examples/shell/main.cpp]: Right, fixed it already. Thank you!

**Review by mspang — APPROVED** [review]

File: examples/shell/cmd_base64.cpp [hunk 2]
```diff
@@ -15,13 +15,13 @@
  *    limitations under the License.
  */
 
-#include <shell/shell.h>
+#include <lib/shell/shell.h>
```
Comment (rwalker-apple) [inline examples/shell/cmd_base64.cpp:18]: this `lib/` element should be in the include directives

File: examples/shell/cmd_device.cpp [hunk 3]
```diff
@@ -15,17 +15,17 @@
...
  */
 
-#include <core/CHIPCore.h>
+#include <lib/core/CHIPCore.h>
 
 #if CONFIG_DEVICE_LAYER
 
 #include <inttypes.h>
 
+#include <lib/shell/shell.h>
```
Comment (rwalker-apple) [inline examples/shell/cmd_device.cpp:24]: same comment as above, the `lib/` element should be in include directives

File: src/lib/core/core.gni [hunk 4]
```diff
@@ -44,7 +44,7 @@ declare_args() {
 }
 
 if (chip_target_style == "") {
-  if (current_os != "freertos") {
+  if (current_os != "freertos" && current_os != "zephyr") {
```
Comment (rwalker-apple) [inline src/lib/core/core.gni:47]: @mspang this would be prettier as an arg?

**Review by rwalker-apple — CHANGES_REQUESTED** [review]
this looks really good.  thanks for all the work here.

would you mind terribly addressing the `lib` review comments?

File: examples/shell/cmd_base64.cpp (hunk 2 above)
Comment (Damian-Nordic) [reply examples/shell/cmd_base64.cpp:18]: I know, however <shell/shell.h> exists in zephyr as well so I had to workaround the file name collision. I guess full paths make such errors less likely to happen.

**Comment by Damian-Nordic** [comment]
> this looks really good. thanks for all the work here.
> 
> would you mind terribly addressing the `lib` review comments?

@rwalker-apple not sure you've seen my comment, but I basically added "lib" component to the include directives to be able to distinguish <shell/shell.h> from CHIP from <shell/shell.h> from Zephyr. Let me know if that's ok or you think it's solveable in a different way.

File: examples/shell/cmd_base64.cpp (hunk 2 above)
Comment (rwalker-apple) [reply examples/shell/cmd_base64.cpp:18]: > > this looks really good. thanks for all the work here.
> > would you mind terribly addressing the `lib` review comments?
> 
> @rwalker-apple not sure you've seen my comment, but I basically added "lib" component to the include directives to be able to distinguish <shell/shell.h> from CHIP from <shell/shell.h> from Zephyr. Let me know if that's ok or you think it's solveable in a different way.

I can't think of a quick fix that's better than yours.  We should think about whether chip's shell needs to be CHIPShell.h.  @turon any ideas?

**Review by rwalker-apple — APPROVED** [review]
@Damian-Nordic will you file an issue about name collisions with Zephyr?

**Review by yhoyoon — APPROVED** [review]

**Pull request merged**